cli-plugins/socket/socket_test.go

```diff
@@ -175,27 +174,25 @@ func TestConnectAndWait(t *testing.T) {
 		}
 	})
 
-	// TODO: this test cannot be executed with `t.Parallel()`, due to
-	// relying on goroutine numbers to ensure correct behaviour
 	t.Run("connect goroutine exits after EOF", func(t *testing.T) {
 		srv, err := NewPluginServer(nil)
 		assert.NilError(t, err, "failed to setup server")
 
 		defer srv.Close()
 
 		t.Setenv(EnvKey, srv.Addr().String())
-		numGoroutines := runtime.NumGoroutine()
 
-		ConnectAndWait(func() {})
-		assert.Equal(t, runtime.NumGoroutine(), numGoroutines+1)
+		ch := make(chan struct{})
+		ConnectAndWait(func() {
+			close(ch)
+		})
 
 		srv.Close()
 
-		poll.WaitOn(t, func(t poll.LogT) poll.Result {
-			if runtime.NumGoroutine() > numGoroutines+1 {
-				return poll.Continue("waiting for connect goroutine to exit")
-			}
-			return poll.Success()
-		}, poll.WithDelay(1*time.Millisecond), poll.WithTimeout(10*time.Millisecond))
```

Comment:
IIRC we specifically wanted to assert that goroutine is stopped.

Wondering if just increasing the timeout would help with the flakiness?

Reply:
i honestly don't know why we would want to do that. The goroutine will stop once the callback function returns.

Reply:
also the failure is not always on the poll, it's on the first assertion on line 190 https://github.com/docker/cli/pull/5927/files/ce9a64c8f4524e6f339f48b1892374c0a5e1580e#diff-2b3ea721470acb354ca37a32dc40b7c0ae55bd43397a084adbd922fca5964e20L190

Reply:
Yeah in this case I think we should also have a poll.WaitOn instead of a straight assert (because the goroutine may not be spawned immediately).

Reply:
> The goroutine will stop once the callback function returns.

Unless there's a bug in ConnectAndWait 🙈

Reply:
but then the test will fail with the timeout...

Reply:
Not really, I'm thinking about a case where the callback is called, but the goroutine is still alive.

Reply:
could you explain how that could happen? It returns after the callback is called here:
https://github.com/docker/cli/blob/ce9a64c8f4524e6f339f48b1892374c0a5e1580e/cli-plugins/socket/socket.go#L164-L165

Reply:
It couldn't happen now, but if `ConnectAndWait` was changed in future, this test would make sure that there's no regression.

Reply:
if the code is changed and the implementation is bad, then there is no guarantee that a test would catch the regression anyway. Focusing on a hypothetical case won't solve the fact that this test fails multiple times per day.

Either we re-design the `ConnectAndWait` to be more resilient or we change the test as I've done here. We cannot predict when a goroutine will run or when it will exit. The changes I make here are more robust from the test perspective.

Reply:
Opened: https://github.com/docker/cli/pull/5932

If it's still flaky after that, I'm 100% agreeing on changing this test according to your proposal :)

Reply:
>But it no longer tests what the test claims to: connect goroutine exits after EOF.

That's not true, it still tests that the goroutine exits on EOF.

The callback is only called once we receive an EOF here:
https://github.com/docker/cli/blob/ce9a64c8f4524e6f339f48b1892374c0a5e1580e/cli-plugins/socket/socket.go#L163-L166

After the callback is called the goroutine will exit due to the `return`.

Reply:
No, it tests that the callback gets called after an EOF.

> After the callback is called the goroutine will exit due to the return.

Yes, that's true with the current code, but tests are also there to detect an unwanted change in future. If this function gets changed in future and introduces a bug where the callback gets called, but there's no `return` out of the goroutine then this test would be able to detect it.